compiler/rustc_mir/src/transform/deduplicate_blocks.rs

```diff
@@ -0,0 +1,233 @@
+//! This pass finds basic blocks that are completely equal,
+//! and replaces all uses with just one of them.
+
+use std::{borrow::Cow, fmt::Debug};
+
+use crate::transform::MirPass;
+
+use itertools::Itertools;
+use rustc_data_structures::fx::FxHashMap;
+use rustc_middle::mir::visit::MutVisitor;
+use rustc_middle::mir::*;
+use rustc_middle::ty::TyCtxt;
+
+use super::simplify::simplify_cfg;
+
+pub struct DeduplicateBlocks {
+    label: String,
+}
+
+impl DeduplicateBlocks {
+    pub fn new(label: &str) -> DeduplicateBlocks {
+        DeduplicateBlocks { label: format!("DeduplicateBlocks-{}", label) }
+    }
+}
+
+impl<'tcx> MirPass<'tcx> for DeduplicateBlocks {
+    fn name(&self) -> Cow<'_, str> {
+        Cow::Borrowed(&self.label)
+    }
+
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        debug!("Running DeduplicateBlocks on `{:?}`", body.source);
+        let duplicates = find_duplicates(body);
+        let should_cleanup = !duplicates.is_empty();
+        let mut opt_applier = OptApplier { tcx, duplicates };
+        opt_applier.visit_body(body);
+
+        if should_cleanup {
+            simplify_cfg(body);
+        }
+    }
+}
+
+fn find_duplicates<'a, 'tcx>(body: &'a Body<'tcx>) -> FxHashMap<BasicBlock, BasicBlock> {
+    let mut duplicates = FxHashMap::default();
+
+    // We want to avoid O(n²) complexity by comparing every basic block with each other.
+    // Basic blocks are equal iff
+    // - their length of statements are equal
+    // - their TerminatorKind are equal
+    // - their StatementKinds are pairwise equal
+    //
+    // The first three cases are cheap to checks, so we use the following strategy:
+    // 1. Sort the basic blocks by (length of statements, TerminatorKind)
+    //      - This is needed for itertools::group_by which only assigns consecutive elements to the same group
+    // 2. Group by (length of statements, TerminatorKind)
+    // 3. Compare StatementKinds pairwise inside the group
+    //      - This is technically O(n²), but `n` should be small in most cases
```

Comment:
If the reason why the [`unicode_normalization` benchmark](https://perf.rust-lang.org/compare.html?start=a1dfd2490a6cb456b92e469fa550dc217e20ad6d&end=50438f458aff5610389abb267c5b5b89fbb68b04) is slow is that `n` is actually large, I wonder if we know why that is? Maybe a large table of constants or something?